File: Programmers/Graph/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <list>
#include <queue>
#include <iostream>
#include <algorithm>
using namespace std;
bool COMPARE(int a, int b)
{
    return a > b;    
}
void bfs(vector<list<int>> &List, vector<int> &count, int n)
{
    int cnt = 0;
    vector<int> visit(n+1,0);
    int temp;
    queue<int> q;
    q.push(1);
    visit[1] = 1; count[1] = 0;
    while(!q.empty())
    {
        temp = q.front();
        q.pop(); 
        cnt++;
        for(auto iter = List[temp].begin();iter != List[temp].end(); iter++)
        {
            if(visit[*iter] != 1)
            {
                q.push(*iter);
                visit[*iter] = 1;
                count[*iter] = count[temp] + 1;
            }
        }
    }
}

int solution(int n, vector<vector<int>> edge) {
    int answer = 1;
    int v1,v2;
    vector<int> count(n+1,0);
    vector<list<int>> List(n + 1);
    
    for(int i = 0; i<edge.size(); i++)
    {
        v1 = edge[i][0];
        v2 = edge[i][1];
        List[v1].push_back(v2);
        List[v2].push_back(v1);
    }
    bfs(List,count, n);
    sort(count.begin(),count.end(),COMPARE);
    int temp = *(count.begin());
    for(auto iter = count.begin() + 1; iter != count.end(); iter++)
    {
        if(temp == *iter)
            answer++;
        else
            break;
    }
    return answer;
}

File: C&Python/Graph/가장 먼 노드.py
import queue

def solution(n, edge):
    # 인접리스트 활용
    answer = 0
    cur_max = 0
    check = [0 for i in range(n + 1)]
    arr = [[0 for i in range(0)] for j in range(n + 1)]
    
    # 양방향 노드의 간선 추가
    for i in range(len(edge)):
        arr[edge[i][0]].append(edge[i][1])
        arr[edge[i][1]].append(edge[i][0])
        
    # 큐 초기화
    q= queue.Queue()
    q.put((1,0))
    check[1] = 1
    
    # bfs -> 마지막 노드 count
    while q.qsize() != 0:
        v, c = q.get()
        if cur_max < c:
            cur_max = c
            answer = 1
        else:
            answer += 1
            
        for i in range(len(arr[v])):
            if check[arr[v][i]] == 0:
                q.put((arr[v][i], c+1))
                check[arr[v][i]] = 1
                
    return answer